src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -53204,67 +53205,73 @@ static void F1(bool b, ref string? x1, ref string y1)
     static void F2(bool b, ref I<string?> x2, ref I<string> y2)
     {
         (b ? ref x2 : ref x2)/*T:I<string?>!*/.P.ToString(); // 6
-        (b ? ref y2 : ref x2)/*T:I<string>!*/.P.ToString(); // 7
-        (b ? ref x2 : ref y2)/*T:I<string>!*/.P.ToString(); // 8, 9
+        (b ? ref y2 : ref x2)/*T:I<string!>!*/.P.ToString(); // 7
+        (b ? ref x2 : ref y2)/*T:I<string!>!*/.P.ToString(); // 8
         (b ? ref y2 : ref y2)/*T:I<string!>!*/.P.ToString();
     }
     static void F3(bool b, ref IIn<string?> x3, ref IIn<string> y3)
     {
         (b ? ref x3 : ref x3)/*T:IIn<string?>!*/.ToString();
-        (b ? ref y3 : ref x3)/*T:IIn<string>!*/.ToString(); // 10
-        (b ? ref x3 : ref y3)/*T:IIn<string>!*/.ToString(); // 11
+        (b ? ref y3 : ref x3)/*T:IIn<string!>!*/.ToString(); // 9
+        (b ? ref x3 : ref y3)/*T:IIn<string!>!*/.ToString(); // 10
         (b ? ref y3 : ref y3)/*T:IIn<string!>!*/.ToString();
     }
     static void F4(bool b, ref IOut<string?> x4, ref IOut<string> y4)
     {
-        (b ? ref x4 : ref x4)/*T:IOut<string?>!*/.P.ToString(); // 12
-        (b ? ref y4 : ref x4)/*T:IOut<string>!*/.P.ToString(); // 13
-        (b ? ref x4 : ref y4)/*T:IOut<string>!*/.P.ToString(); // 14
+        (b ? ref x4 : ref x4)/*T:IOut<string?>!*/.P.ToString(); // 11
+        (b ? ref y4 : ref x4)/*T:IOut<string?>!*/.P.ToString(); // 12, 13
```

Comment:
I compared this test to an equivalent test with type inference and I'm seeing different warnings. It would be good to understand why and whether it is unexpected. My expectation is that they would match.

```
interface I<T> { T P { get; } }
interface IIn<in T> { }
interface IOut<out T> { T P { get; } }
class C
{
    static void F1(bool b, ref string? x1, ref string y1)
    {
        M(ref x1, ref x1).ToString(); // 1 on M
        M(ref x1, ref y1).ToString(); // 2 on x1
        M(ref y1, ref x1).ToString();
        M(ref y1, ref y1).ToString();
    }
    static void F2(bool b, ref I<string?> x2, ref I<string> y2)
    {
        M(ref x2, ref x2).P.ToString(); // 6 on P
        M(ref y2, ref x2).P.ToString(); // 7 on x2
        M(ref x2, ref y2).P.ToString(); // 8 on x2
        M(ref y2, ref y2).P.ToString();
    }
    static void F3(bool b, ref IIn<string?> x3, ref IIn<string> y3)
    {
        M(ref x3, ref x3).ToString();
        M(ref y3, ref x3).ToString(); // 9 on x3
        M(ref x3, ref y3).ToString(); // 10 on x3
        M(ref y3, ref y3).ToString();
    }
    static void F4(bool b, ref IOut<string?> x4, ref IOut<string> y4)
    {
        M(ref x4, ref x4).P.ToString(); // 11 on M
        M(ref y4, ref x4).P.ToString(); // 12 on x4
        M(ref x4, ref y4).P.ToString(); // 14 on x4
        M(ref y4, ref y4).P.ToString();
    }
    static T M<T>(ref T t1, ref T t2) => throw null!;
}
```
[sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCecAmIA1AD6wIZYQDGqABAJIA8AKgHx0DedLdAClzoBzVAgDcdAL5SAsAChylGvQYMYTWDw7dJ8xeiq1GAeQCuCJgHtzWwbwHcR4qbLkABAEx0AwvM7y6QLo3AEYANmCAFjoAMRCACmBLSwAbOmA4OnRULGCQgAYAfjoADxDM7NzQ/LocEIBKAKD/OSC2ugBZeMrS8qyc3vqAOhZLAGUETBghePqJAHp5uhC6SxhOpvbO7oGyioG64dGJqZm5ukW6LzXezfaunrr93LKj8cnYM7E7toeDvseDRG71Os2+rSCughgVCETc0RiHkSyTSGX6uWY1UKHBKHmejCY1Q4OA8jWhXB+QT+Lzx6NKpKGfGBJ0+YIuSwiNz4lMC1NqtJ6uOGTOOH2mbMuAHZVutcTztoKBQcGSKQazzpcABwy+nyvkk/Ek4XMsVfO5QtqwqKxADMyNS6XxqnUWJxNqdakJBWJNrJbRaW15Oxe7rpJV9JtBcz1wdqocFEdFUYWSwAnDrwzGE4bE2rxRqlgUMzaswd42W3iz8+C2haglb4bFIvbUU6zBZXaVIm3zF78sTIn7mqWXt2w4PGZH1Snlisbh0R7Ux4KJ6qq192ctrrLIouSsuDqup9XNyFojd94ucAfctfjUnp+a7lbeB1WGxY7wEACBl/SXQAF4OAQAALdBLAAdzoGBTBSFIAEJwUkIA)

Reply:
That's related to the previous discussion. When there is a mismatch (`ref string` vs `ref string?` arguments), type inference infers the method as `M<string>`. But for the ternary, we combine the annotations via `EnsureCompatible`, i.e., it becomes `ref string?`. The flow state might be also different for similar reasons.

I'm not sure we should strive for precise equivalence between type inference and the ternary. The ternary is a conditional construct whose branches are joined together which the method call is not.

Btw, I don't see any difference on this line (there is no warning either way):
> M(ref x3, ref x3).ToString(); // (difference)

Reply:
It took me a while to understand, but I think you're correct. The difference in warnings is expected.

Here's what I think is happening for the following ref ternary scenario:
```
static void F1(bool b, ref string? x1, ref string y1)
{
    (b ? ref x1 : ref y1)/*T:string?*/.ToString(); // 2 on x1, 3 on dereference
}
```
The type for the ternary is `string!` (result from `EnsureCompatible` between `!` and `?`).
But when we analyze the dereference, we look at the state (maybe-null) so we warn on the dereference.

nit: Consider adding scenarios showing that result type. For example, doing a ref assignment to the ternary `(b ? ref x1 : ref y1) = ref ...;`.

In contrast for the method inference scenario:
```
static void F1(bool b, ref string? x1, ref string y1)
{
    M(ref x1, ref y1).ToString(); // 2 on x1 (which is correct in my opinion)
}
```
We also produce the same result type (`string!`, from `EnsureCompatible`) but we don't have a tracked state so we just derive a state from the type (not-null) so there is no warning.

In short, I think the logic is good (we infer the same best type in both cases).

> The ternary is a conditional construct whose branches are joined together which the method call is not.

This comment made me think of another scenario.

```
#nullable enable 

var x = (bool b, ref object o1, ref object? o2) =>
{
    if (b)
    {
        return ref o1;
    }
    return ref o2; // bug: missing warning
};

object x1 = new object();
object? x2 = new object();
bool b = true;

x(b, ref x1, ref x2) = null; // bug: missing warning (inferred delegate should return `ref string!` but actually returns `ref string~`)

ref object? y1 = ref x(b, ref x1, ref x2); // bug: missing warning (inferred delegate should return `ref string!` but actually returns `ref string~`)

ref object y2 = ref x(b, ref x1, ref x2);
```

I would expect that scenario (multiple ref-returns) to behave the same as type inference with multiple ref arguments and as ref ternaries. The return type from `ref string?` and `ref string!` should be `ref string!`.
Unfortunately, there seems to be a bug with `var` local analysis (we're using an anonymous delegate type from initial binding, with `objet~` return) so the current implementation is not very useful at the moment for making comparisons.
Filed https://github.com/dotnet/roslyn/issues/75009